Onboardingtask

kmeena01 opened the pull request.

kmeena01 committed:
Onboardingtask

kmeena01 committed:
Testcase

kmeena01 committed:
I changed for quiting chrome and clear existing data before doing some steps

kmeena01 committed:
missed before

amanbansalnz commented on Features/Addlanguage.feature:
```diff
@@ -0,0 +1,57 @@
+	Given User logs into Mars Portal 
+	When clean the existing data and add new language 'Malay' 
+	Then new Language 'Malay' Updated successfully
+
```
Please increase your coverage around the features you  are testing here is an idea of what is missing 

• Negative testing with valid input
• Negative testing with invalid input
• Destructive testing

E.g of above scenarios
Happy path tests check basic functionality and the acceptance criteria of the API. We later extend positive tests to include optional parameters and extra functionality.

The next group of tests is negative testing where we expect the application to gracefully handle problem scenarios with both valid user input (for example, trying to add an existing username) and invalid user input (trying to add a username which is null).

Destructive testing is a deeper form of negative testing where we intentionally attempt to break the API to check its robustness (for example, sending a huge payload body in an attempt to overflow the system).

amanbansalnz commented on Features/Addskills.feature:
```diff
@@ -0,0 +1,76 @@
+@tag1
+Scenario:1. Verify User able to Add New Skill
+	Given User logging in Mars portal 
+	And User navigates to Skills tab
```
same comments apply

amanbansalnz commented on StepDefinitions/AddlangStepDefinitions.cs:
```diff
@@ -0,0 +1,124 @@
+    [Binding]
+    public class AddlangStepDefinitions
+    {
+        IWebDriver driver = new ChromeDriver();
```
Please make use of constructors, please see 
 https://www.tutorialspoint.com/csharp/csharp_classes.htm

https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

amanbansalnz commented on StepDefinitions/AddlangStepDefinitions.cs:
```diff
@@ -0,0 +1,124 @@
+        [When(@"User creates a new Language record '([^']*)'")]
+        public void WhenUserCreatesANewLanguageRecord(string language)
+        {
+            Addmultiplelanguageobj.cleardata(driver);
```
so Addmultiplelanguageobj.cleardata(driver); needs to be put in a hook ( one that runs before any test), and then at the end of this test you should only clean up the data this tests adds

amanbansalnz commented on StepDefinitions/AddlangStepDefinitions.cs:
```diff
@@ -0,0 +1,124 @@
+        [Then(@"new Language '([^']*)' should be successfully created")]
+        public void ThenNewLanguageShouldBeSuccessfullyCreated(string language)
+        {
+            Addmultiplelanguageobj.verifyaddedinlist(driver, language);
```
Please do not pass the driver In action methods (methods which in pages doing things like create, delete etc ) instead get the pages to extend to common driver where you will create a common driver class which will initialise the web driver 

public class CommonDriver
    {
        public static IWebDriver driver;

        public void initialise()
        {
            // create driver
            driver = new ChromeDriver();

        }

    }

amanbansalnz replied:
fix this everywhere you have done this

amanbansalnz commented on Utilities/CommonDriver.cs:
```diff
@@ -0,0 +1,14 @@
+namespace SpecProj2.Utilities
+{
+    public class CommonDriver
+    {
```
finish implementing this

amanbansalnz replied:
please make use of Hooks , you might need to create the Common driver object there!
Please see
https://codebun.com/how-to-implement-hooks-in-specflow-c/

amanbansalnz commented on StepDefinitions/AddskillsStepDefinitions.cs:
```diff
@@ -0,0 +1,163 @@
+    public class AddskillsStepDefinitions 
+    {
+        IWebDriver driver = new ChromeDriver();
+        Login Loginobj = new Login();
```
same comments apply

amanbansalnz commented on Pages/AddNewSkill.cs:
```diff
@@ -0,0 +1,146 @@
+
+namespace SpecProj2.Pages
+{
+    public class AddNewSkill
```
you should create three pages Login Page, Skill Page and Language Page

amanbansalnz replied:
Please implement the POM Pattern, please see the below 
https://docs.specflow.org/projects/specflow/en/latest/ui-automation/Selenium-with-Page-Object-Pattern.html

amanbansalnz replied:
same comments apply to all pages

amanbansalnz commented on Pages/AddNewSkill.cs:
```diff
@@ -0,0 +1,146 @@
+            wait.Until(SeleniumExtras.WaitHelpers.ExpectedConditions.ElementExists(By.XPath("//div[@class='ns-box-inner']")));
+            IWebElement errorMessage = driver.FindElement(By.XPath("//div[@class='ns-box-inner']"));
+            string errorMessageText = errorMessage.Text;
+            Assert.That(errorMessage.Text == "This skill is already exist in your skill list.", "Already exists");
```
Asserts should live in steps please move all asserts there

amanbansalnz reviewed (changes requested):
I have reviewed your task  please make the following changes which is suggested in the pr

kmeena01 committed:
For the comment of Negative testing - I hope added some negative testcases

For the comment of  Addmultiplelanguageobj.cleardata(driver); needs to be put in a hook ( one that runs before any test), and then at the end of this test you should only clean up the data this tests adds  - I cannot cleardata for all testcases because if it cleanup then cannot cover duplicate data scenario so i add on wherever needed.

For the comment of Hook and commondriver - done added and implemented

For the comment of you should create three pages Login Page, Skill Page and Language Page - Done

kmeena01 committed:
For the comment of Negative testing - I hope added some negative testcases

For the comment of  Addmultiplelanguageobj.cleardata(driver); needs to be put in a hook ( one that runs before any test), and then at the end of this test you should only clean up the data this tests adds  - I cannot cleardata for all testcases because if it cleanup then cannot cover duplicate data scenario so i add on wherever needed.

For the comment of Hook and commondriver - done added and implemented

For the comment of you should create three pages Login Page, Skill Page and Language Page - Done

amanbansalnz commented on StepDefinitions/AddlangStepDefinitions.cs:
```diff
@@ -0,0 +1,139 @@
+    [Binding]
+    public class AddlangStepDefinitions : CommonDriver
+    {
+        Login Loginobj = new Login();
```
Please make use of constructors, please see 
 https://www.tutorialspoint.com/csharp/csharp_classes.htm

https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

amanbansalnz commented on Utilities/Hook.cs:
```diff
@@ -0,0 +1,28 @@
+    public class Hook : CommonDriver
+    {
+        
+        [Before]
```
are these hooks correct, https://codebun.com/how-to-implement-hooks-in-specflow-c/ please use the correct specflow hooks

amanbansalnz commented on Pages/Addlanguage.cs:
```diff
@@ -0,0 +1,169 @@
+{
+    public class Addlanguage : CommonDriver
+    {
+        public void addlanguage(string language, string level)
```
you still not have implemented the POM pattern 

see https://docs.specflow.org/projects/specflow/en/latest/ui-automation/Selenium-with-Page-Object-Pattern.html

amanbansalnz commented on StepDefinitions/AddlangStepDefinitions.cs:
```diff
@@ -0,0 +1,139 @@
+        [When(@"User creates a new Language record '([^']*)' '([^']*)'")]
+        public void WhenUserCreatesANewLanguageRecord(string language, string level)
+        {
+            Addlanguageobj.cleardata();
```
you should clear all data, only once  before any tests start, and then clean up test data only which the test generates
